### PR TITLE
Fix TA cache maintenance support

### PR DESCRIPTION
**Pull request by pascal-brand-st-dev**
Fix the 2-level cache clean/inv/flush sequences.
Current implementation follows recommendations from
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0246d/Beicdhde.html

Tested-by: Etienne CARRIERE <etienne.carriere@st.com>
Reviewed-by: CIBUILD <smet-aci-builds@lists.codex.cro.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@st.com>
Signed-off-by: Pascal Brand <pascal.brand@st.com>

### CODE REVIEW

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 1]
```diff
@@ -1614,27 +1614,40 @@ TEE_Result tee_uta_cache_operation(struct tee_ta_session *sess,
 		return TEE_ERROR_ACCESS_DENIED;
 
 	switch (op) {
+
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: Empty line not needed.

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 2]
```diff
@@ -1614,27 +1614,40 @@ TEE_Result tee_uta_cache_operation(struct tee_ta_session *sess,
...
 	case TEE_CACHEFLUSH:
-		l1op = DCACHE_AREA_CLEAN_INV;
-		l2op = L2CACHE_AREA_CLEAN_INV;
-		break;
+		/* clean L1, dsb, flush L2, inval L1, dsb. */
+		ret = cache_maintenance_l1(DCACHE_AREA_CLEAN, va, len);
+		if (ret != TEE_SUCCESS)
+			return ret;
+		dsb();
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: cache_maintenance_l1() already does a dsb() before returning.

**Comment by pascal-brand-st-dev** [comment]
Review applied

File: core/arch/arm/kernel/tee_ta_manager.c [hunk 3]
```diff
@@ -1615,26 +1615,34 @@ TEE_Result tee_uta_cache_operation(struct tee_ta_session *sess,
 
 	switch (op) {
 	case TEE_CACHEFLUSH:
-		l1op = DCACHE_AREA_CLEAN_INV;
-		l2op = L2CACHE_AREA_CLEAN_INV;
-		break;
+		/* clean L1, dsb, flush L2, inval L1, dsb. */
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/tee_ta_manager.c]: The comment isn't consistent with the code, both with respect to dsb and the final inval l1.

The ARM document recommended an inval l1 at the end of the sequence. Clean-inval l1 as is done here seems safer though, especially as we don't guarantee that there's no accesses to the memory during these operations.

**Comment by pascal-brand-st-dev** [comment]
Applied comments

**Comment by jenswi-linaro** [comment]
```
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

**Pull request merged**